### ISSUE TITLE
[MNG-5605] ssh-wagon hangs

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6704]
**[Frank Cornelis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fcorneli)** opened **[MNG-5605](https://issues.apache.org/jira/browse/MNG-5605?redirect=false)** and commented

When releasing (using maven-release-plugin) via Maven 3.1.1 everything works as expected. When doing the same via Maven 3.2.1, ssh-wagon all of the sudden hangs on the second ssh upload.

---

**Affects:** 3.2.1

**Issue Links:**
- [WAGON-429](https://issues.apache.org/jira/browse/WAGON-429) Scp download hangs
 (_**"depends upon"**_)

5 votes, 16 watchers

**Comment by jira-importer** [comment]
**[Allan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=as987)** commented

This happens consistently for me when deploying a file.

OS: Windows 7 64 bit
JVM: Oracle JDK 6u45 64 bit
Debug console output:

```
...
[DEBUG] Configuring mojo org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy from plugin realm ClassRealm[plugin>org.apache.maven.plugins:maven-deploy-plugin:2.7, parent: sun.misc.Launcher$AppClassLoader@546b97fd]
[DEBUG] Configuring mojo 'org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy' with basic configurator -->
[DEBUG]   (f) artifact = my-group:my-project:jar:1.0-SNAPSHOT
[DEBUG]   (f) attachedArtifacts = [my-group:my-project:java-source:sources:1.0-SNAPSHOT, my-group:my-project:java-source:test-sources:1.0-SNAPSHOT]
[DEBUG]   (s) localRepository =        id: local
      url: file:///C:/Users/username/.m2/repository/
   layout: default
snapshots: [enabled => true, update => always]
 releases: [enabled => true, update => always]

[DEBUG]   (f) offline = false
[DEBUG]   (f) packaging = jar
[DEBUG]   (f) pomFile = C:\my-group\my-project\pom.xml
[DEBUG]   (f) project = MavenProject: my-group:my-project:1.0-SNAPSHOT @ C:\my-group\my-project\pom.xml
[DEBUG]   (f) retryFailedDeploymentCount = 1
[DEBUG]   (f) skip = false
[DEBUG]   (f) updateReleaseInfo = false
[DEBUG] -- end configuration --
[DEBUG] Using connector WagonRepositoryConnector with priority 0.0 for scp://repo.example.com/snapshot-repo/ with username=username, password=***
Downloading: scp://repo.example.com/snapshot-repo/com/example/my-group/my-project/1.0-SNAPSHOT/maven-metadata.xml
2/2 KB
```

Thread dump during hang:

```
2014-03-25 10:58:02
Full thread dump Java HotSpot(TM) 64-Bit Server VM (20.45-b01 mixed mode):

"Connect thread repo.example.com session" prio=6 tid=0x000000000ae81000 nid=0x16a4 runnable [0x000000000cf4f000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:129)
	at com.jcraft.jsch.IO.getByte(IO.java:82)
	at com.jcraft.jsch.Session.read(Session.java:908)
	at com.jcraft.jsch.Session.run(Session.java:1378)
	at java.lang.Thread.run(Thread.java:662)

"pool-1-thread-1" prio=6 tid=0x000000000ae06000 nid=0x187c in Object.wait() [0x000000000a7df000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000007af0ddb90> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at java.io.PipedInputStream.read(PipedInputStream.java:310)
	- locked <0x00000007af0ddb90> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at java.io.PipedInputStream.read(PipedInputStream.java:361)
	- locked <0x00000007af0ddb90> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at org.apache.maven.wagon.AbstractWagon.transfer(AbstractWagon.java:532)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:354)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:317)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:286)
	at org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:97)
	at org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61)
	at org.eclipse.aether.connector.wagon.WagonRepositoryConnector$GetTask.run(WagonRepositoryConnector.java:660)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:67)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)

"Low Memory Detector" daemon prio=6 tid=0x0000000007834000 nid=0x2d48 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread1" daemon prio=10 tid=0x0000000007827000 nid=0x2a7c waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" daemon prio=10 tid=0x0000000007820000 nid=0x1cb4 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Attach Listener" daemon prio=10 tid=0x000000000781e800 nid=0x28fc waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Signal Dispatcher" daemon prio=10 tid=0x000000000781c000 nid=0x1f60 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Finalizer" daemon prio=8 tid=0x00000000077c4000 nid=0x2bb8 in Object.wait() [0x000000000850f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x0000000707123230> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:118)
	- locked <0x0000000707123230> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:134)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:171)

"Reference Handler" daemon prio=10 tid=0x00000000077c2800 nid=0x18c8 in Object.wait() [0x000000000840f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x0000000707123260> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:485)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)
	- locked <0x0000000707123260> (a java.lang.ref.Reference$Lock)

"main" prio=6 tid=0x000000000057c800 nid=0x2590 waiting on condition [0x0000000002fee000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:283)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder.awaitTerminationOfAllRunnables(RunnableErrorForwarder.java:129)
	at org.eclipse.aether.util.concurrency.RunnableErrorForwarder.await(RunnableErrorForwarder.java:96)
	at org.eclipse.aether.connector.wagon.WagonRepositoryConnector.get(WagonRepositoryConnector.java:503)
	at org.eclipse.aether.internal.impl.DefaultDeployer.upload(DefaultDeployer.java:433)
	at org.eclipse.aether.internal.impl.DefaultDeployer.deploy(DefaultDeployer.java:313)
	at org.eclipse.aether.internal.impl.DefaultDeployer.deploy(DefaultDeployer.java:268)
	at org.eclipse.aether.internal.impl.DefaultRepositorySystem.deploy(DefaultRepositorySystem.java:413)
	at org.apache.maven.artifact.deployer.DefaultArtifactDeployer.deploy(DefaultArtifactDeployer.java:139)
	at org.apache.maven.plugin.deploy.AbstractDeployMojo.deploy(AbstractDeployMojo.java:167)
	at org.apache.maven.plugin.deploy.DeployMojo.execute(DeployMojo.java:157)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:108)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:76)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:116)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:361)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)

"VM Thread" prio=10 tid=0x00000000077b9000 nid=0x2a20 runnable 

"GC task thread#0 (ParallelGC)" prio=6 tid=0x00000000004a8000 nid=0x2f1c runnable 

"GC task thread#1 (ParallelGC)" prio=6 tid=0x00000000004a9800 nid=0x1874 runnable 

"GC task thread#2 (ParallelGC)" prio=6 tid=0x00000000004ab800 nid=0x23f8 runnable 

"GC task thread#3 (ParallelGC)" prio=6 tid=0x00000000004ad800 nid=0x1b1c runnable 

"GC task thread#4 (ParallelGC)" prio=6 tid=0x00000000004b1000 nid=0x236c runnable 

"GC task thread#5 (ParallelGC)" prio=6 tid=0x00000000004b2000 nid=0x2834 runnable 

"GC task thread#6 (ParallelGC)" prio=6 tid=0x00000000004b3800 nid=0x2d3c runnable 

"GC task thread#7 (ParallelGC)" prio=6 tid=0x00000000004b5000 nid=0x2eec runnable 

"VM Periodic Task Thread" prio=10 tid=0x0000000007846000 nid=0x2ac8 waiting on condition 

JNI global references: 1806
```

When using maven 3.1.1, it successfully completes.

**Comment by jira-importer** [comment]
**[Parolini Antonio](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=aparolini)** commented

Same here. I tried to upgrade to the latest wagon ssh (2.6), same bug using maven 3.2.1. Fine with maven 3.1.1, thus this looks a maven core issue.

Will be great if this is corrected in next maven release.

**Comment by jira-importer** [comment]
**[Gray Watson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=graywatson)** commented

I'm having a similar problem with maven downloads that hang -- typically on very small files like poms.   If I log into the repo box I see the following process:

> `maven    20954  0.0  0.0   2388   736 ?        Ss   16:45   0:00 scp -p -f /var/maven/org/moxieapps/gwt-highcharts/1.5.0/gwt-highcharts-1.5.0.pom`

If I then kill the process on the server, the client continues without an error.

> `Downloading: scp://dev.achievementnetwork.org/var/maven/org/moxieapps/gwt-highcharts/1.5.0/gwt-highcharts-1.5.0.pom`
> `401/400 B`

It seems to me to be very small files that have the problem.  Possibly smaller than some sort of packet?  2k, 400B, etc..

**Comment by jira-importer** [comment]
**[Gray Watson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=graywatson)** commented

I seemed to have worked around this problem by switching to use sftp URLs as opposed to using scp.  In looking at the code, it sure seems to be that this is some incompatibility issue between jsch and wagon-ssh.

**Comment by jira-importer** [comment]
**[Marco Brandizi](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=zakmck)** commented

sftp worked for me too. Thank you so much! :-)
Does this mean I'm no longer using wagon-ssh and instead I'm using jsch? Sorry, I'm too unfamiliar with this part of Maven to know if I'm asking a silly thing ;-)

**Comment by jira-importer** [comment]
**[Gray Watson](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=graywatson)** commented

No you are still using wagon-ssh _and_ jsch, just that is is using the jsch sftp protocol implementation instead of the scp protocol.  Not sure why sftp works.

**Comment by jira-importer** [comment]
**[Arnaud Brunet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gronono)** commented

Using wagon-ssh 2.6 and maven 3.2.1, I have this issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I'll reopen until some one verifies this as fixed with Maven 3.2.2.

**Comment by jira-importer** [comment]
**[Frank Cornelis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fcorneli)** commented

The problem still occurs on Maven 3.2.2.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Has anyone tried to downgrade/upgrade Wagon SSH or another JSch version? I doubt that this is a Maven problem but solely a plugin problem. I could be wrong, though.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Anyone with the issue needs to try Maven 3.2.2 with Wagon 2.4 which is what 3.1.1 uses. None of the resolver code has changed as we are still at Aether 0.0.9.M2.

It can't be fixed in Maven 3.2.2 if it's a Wagon issue.

**Comment by jira-importer** [comment]
**[Allan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=as987)** commented

For me, at least, this does not seem to be a wagon-ssh issue. With the wagon-ssh version locked, I can recreate this issue 100% of the time simply by flipping between 3.2.1 (broken) and 3.1.1 (working). I have tried this with both wagon-ssh versions 2.4 and 2.6 with the same result.

Unfortunately, I can't seem to test Maven 3.2.2 - I'm running into the same problems that were discussed on the mailing list in [this thread](http://mail-archives.apache.org/mod_mbox/maven-dev/201407.mbox/%3C50066049-F6D1-4BF2-8C12-577658F311FA@takari.io%3E)

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Did how did you use the different versions? Did you also make sure you were using the different versions of jsch? Wagon 2.4 uses 0.1.44-1, while Wagon 2.6 uses 0.1.50.

**Comment by jira-importer** [comment]
**[Allan](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=as987)** commented

I assumed that locking the version in the extension declaration would use the same version regardless of which version of maven I'm using.

```xml
<build>
...
    <extensions>
        <extension>
            <groupId>org.apache.maven.wagon</groupId>
            <artifactId>wagon-ssh</artifactId>
            <version>2.6</version>
        </extension>
</build>
```

Is that not the case?

Also, I didn't do anything else to control the jsch version.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

That should, I didn't know if you used it as an extension or dropped the JARs in your installation. Nothing else around transport has changed, most notably the Aether code where the error is originating. Honestly this is not high on my priority list (which is not to say there aren't other committers for whom this is important)  as I don't use the SSH code path. Ideally if we can get a self-contained test that uses something like MINA SSHD and can demontrate the problem it will be easier to fix, and more importantly stays fixed.

**Comment by jira-importer** [comment]
**[Vladimir Piyanov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=vpiyanov)** commented

I ran into the same problem while using the **download-single** goal and maven 3.2.2.

```
<build>
    <extensions>
        <extension>
            <groupId>org.apache.maven.wagon</groupId>
            <artifactId>wagon-ssh</artifactId>
            <version>2.6</version>
        </extension>
    </extensions>
</build>
...
                    <plugin>
                        <groupId>org.codehaus.mojo</groupId>
                        <artifactId>wagon-maven-plugin</artifactId>
                        <version>1.0-beta-5</version>
                        <executions>
                            <execution>
                                <id>2-download-dump</id>
                                <phase>install</phase>
                                <goals>
                                    <goal>download-single</goal>
                                </goals>
                                <configuration>
                                    <url>scp://${jdbc.ssh.username}:${jdbc.ssh.password}@${jdbc.ssh.host}:${jdbc.ssh.port}/</url>
                                    <fromFile>${jdbc.ssh.dump.dir.path}/${jdbc.ssh.dump.file.name}</fromFile>
                                    <toFile>${basedir}/${jdbc.ssh.dump.file.name}</toFile>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
```

Thread dump:

```

2014-08-13 12:58:05
Full thread dump Java HotSpot(TM) 64-Bit Server VM (20.45-b01 mixed mode):

"RMI TCP Connection(2)-10.99.66.53" daemon prio=6 tid=0x00000000094e5800 nid=0xa98 runnable [0x000000000babf000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:129)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:237)
	- locked <0x00000000edaf0e30> (a java.io.BufferedInputStream)
	at java.io.FilterInputStream.read(FilterInputStream.java:66)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:517)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:790)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:649)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- <0x00000000ed7234b0> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)

"JMX server connection timeout 15" daemon prio=6 tid=0x0000000007b8b000 nid=0x1d68 in Object.wait() [0x000000000b9bf000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ed811580> (a [I)
	at com.sun.jmx.remote.internal.ServerCommunicatorAdmin$Timeout.run(ServerCommunicatorAdmin.java:150)
	- locked <0x00000000ed811580> (a [I)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- None

"RMI Scheduler(0)" daemon prio=6 tid=0x0000000008e60800 nid=0x2154 waiting on condition [0x000000000b8bf000]
   java.lang.Thread.State: TIMED_WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000ed5e2710> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:196)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2025)
	at java.util.concurrent.DelayQueue.take(DelayQueue.java:164)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:609)
	at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:602)
	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:957)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:917)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- None

"RMI TCP Connection(1)-10.99.66.53" daemon prio=6 tid=0x0000000008885800 nid=0x21f0 runnable [0x000000000b7bf000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:129)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:218)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:237)
	- locked <0x00000000ed7bd9e0> (a java.io.BufferedInputStream)
	at java.io.FilterInputStream.read(FilterInputStream.java:66)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:517)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:790)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:649)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- <0x00000000ed720ff0> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)

"RMI TCP Accept-0" daemon prio=6 tid=0x0000000007e59000 nid=0x1c70 runnable [0x000000000b6bf000]
   java.lang.Thread.State: RUNNABLE
	at java.net.PlainSocketImpl.socketAccept(Native Method)
	at java.net.PlainSocketImpl.accept(PlainSocketImpl.java:408)
	- locked <0x00000000ed5e9888> (a java.net.SocksSocketImpl)
	at java.net.ServerSocket.implAccept(ServerSocket.java:462)
	at java.net.ServerSocket.accept(ServerSocket.java:430)
	at sun.management.jmxremote.LocalRMIServerSocketFactory$1.accept(LocalRMIServerSocketFactory.java:34)
	at sun.rmi.transport.tcp.TCPTransport$AcceptLoop.executeAcceptLoop(TCPTransport.java:369)
	at sun.rmi.transport.tcp.TCPTransport$AcceptLoop.run(TCPTransport.java:341)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- None

"Connect thread 192.168.56.217 session" prio=6 tid=0x0000000008bf3800 nid=0x1660 runnable [0x000000000a3ef000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:129)
	at com.jcraft.jsch.IO.getByte(IO.java:82)
	at com.jcraft.jsch.Session.read(Session.java:821)
	at com.jcraft.jsch.Session.run(Session.java:1241)
	at java.lang.Thread.run(Thread.java:662)

   Locked ownable synchronizers:
	- None

"Low Memory Detector" daemon prio=6 tid=0x0000000006901000 nid=0x1b30 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread1" daemon prio=10 tid=0x00000000068ff800 nid=0x1fd4 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread0" daemon prio=10 tid=0x00000000068fa800 nid=0x1410 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Attach Listener" daemon prio=10 tid=0x00000000068f8800 nid=0x13c4 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Signal Dispatcher" daemon prio=10 tid=0x00000000068ef800 nid=0x764 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Finalizer" daemon prio=8 tid=0x00000000002c6800 nid=0x984 in Object.wait() [0x000000000689f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ef550348> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:118)
	- locked <0x00000000ef550348> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:134)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:171)

   Locked ownable synchronizers:
	- None

"Reference Handler" daemon prio=10 tid=0x00000000002c4800 nid=0x2094 in Object.wait() [0x000000000679f000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000ef5503e8> (a java.lang.ref.Reference$Lock)
	at java.lang.Object.wait(Object.java:485)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:116)
	- locked <0x00000000ef5503e8> (a java.lang.ref.Reference$Lock)

   Locked ownable synchronizers:
	- None

"main" prio=6 tid=0x000000000035b800 nid=0xafc in Object.wait() [0x00000000025de000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000000eaefa960> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at java.io.PipedInputStream.read(PipedInputStream.java:310)
	- locked <0x00000000eaefa960> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at java.io.PipedInputStream.read(PipedInputStream.java:361)
	- locked <0x00000000eaefa960> (a com.jcraft.jsch.Channel$MyPipedInputStream)
	at org.apache.maven.wagon.AbstractWagon.transfer(AbstractWagon.java:532)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:354)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:317)
	at org.apache.maven.wagon.AbstractWagon.getTransfer(AbstractWagon.java:286)
	at org.apache.maven.wagon.StreamWagon.getIfNewer(StreamWagon.java:97)
	at org.apache.maven.wagon.StreamWagon.get(StreamWagon.java:61)
	at org.codehaus.mojo.wagon.shared.DefaultWagonDownload.download(DefaultWagonDownload.java:87)
	at org.codehaus.mojo.wagon.DownloadMojo.execute(DownloadMojo.java:51)
	at org.codehaus.mojo.wagon.AbstractSingleWagonMojo.execute(AbstractSingleWagonMojo.java:64)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:347)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:154)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:213)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:157)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:597)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)

   Locked ownable synchronizers:
	- None

"VM Thread" prio=10 tid=0x00000000002bc800 nid=0x21fc runnable 

"GC task thread#0 (ParallelGC)" prio=6 tid=0x0000000000217000 nid=0xe68 runnable 

"GC task thread#1 (ParallelGC)" prio=6 tid=0x0000000000218800 nid=0x1084 runnable 

"GC task thread#2 (ParallelGC)" prio=6 tid=0x000000000021a800 nid=0x960 runnable 

"GC task thread#3 (ParallelGC)" prio=6 tid=0x000000000021e800 nid=0x181c runnable 

"VM Periodic Task Thread" prio=10 tid=0x0000000006913800 nid=0x1ca0 waiting on condition 

JNI global references: 1088


```

**Comment by jira-importer** [comment]
**[Mariela Daventini](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mdaventini)** commented

Affects version 3.2.3

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=danttran) can this be caused by WAGON-431?

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=danttran)** commented

I am very convinced this is related to WAGON-429 where the scp download accidentally set the requested transfer size as Long.MAX_VALUE rather than the size of the remote file.

See my explaination at WAGON-429 for details

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

You can download one of the latest SNAPSHOTs of the Maven distribution from https://repository.apache.org/content/repositories/snapshots/org/apache/maven/apache-maven/3.2.6-SNAPSHOT/ and give it a try. It relies on the latest SNAPSHOT of Wagon, which should contain the required fixes.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

can someone confirm that the SNAPSHOT wagon fixes the issue, please?
to be sure releaing Wagon (and waiting for it before releasing Maven core) is the way to go

thanks in advance to everybody interested in seeing this issue fixed

**Comment by jira-importer** [comment]
**[Frank Cornelis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fcorneli)** commented

I'm trying out 3.2.6-SNAPSHOT with the project at: https://github.com/e-Contract/mycarenet (check the pom.xml for the used release config).
However, during `mvn release:perform` I still get:

```
[INFO] [INFO] --- maven-deploy-plugin:2.8.2:deploy (default-deploy) @ mycarenet ---
[INFO] Uploading: scp://192.168.100.1/maven2/be/e_contract/mycarenet/1.2.1/mycarenet-1.2.1.pom
[INFO] 4/8 KB   
[INFO] 8/8 KB   
[INFO]          
[INFO] Uploaded: scp://192.168.100.1/maven2/be/e_contract/mycarenet/1.2.1/mycarenet-1.2.1.pom (8 KB at 4.2 KB/sec)
[INFO] Downloading: scp://192.168.100.1/maven2/be/e_contract/mycarenet/maven-metadata.xml
[INFO] 892/891 B
```

The reported downloading size is one off and simply hangs on that.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I realized a bit too late that wagon-ssh isn't part of the Maven distribution, see http://maven.apache.org/ref/3.2.5/apache-maven/dependencies.html
However, it surprises me that this used to work with Maven-3.1.x and not in Maven-3.2.x anymore, like there's some kind of classloader issue. At least give wagon-ssh 2.9-SNAPSHOT a try, I assume you have specified wagon-ssh somewhere.

**Comment by jira-importer** [comment]
**[Frank Cornelis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fcorneli)** commented

Of course the maven-release-plugin refuses to prepare for a release when depending on a wagon-ssh 2.9-SNAPSHOT version. So I guess I'll have to do a local "release" of wagon-ssh first.

**Comment by jira-importer** [comment]
**[Frank Cornelis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fcorneli)** commented

Doing a quick-and-dirty local "release" of wagon-ssh via:

```
git clone https://git-wip-us.apache.org/repos/asf/maven-wagon.git
cd maven-wagon
mvn versions:set -DnewVersion=2.9-econtract-1
mvn clean install -Dmaven.test.skip=true
```

and then referring to it via:

```xml
<extensions>
    <extension>
        <groupId>org.apache.maven.wagon</groupId>
        <artifactId>wagon-ssh</artifactId>
        <version>2.9-econtract-1</version>
    </extension>
</extensions>
```

indeed fixes the issue.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

When doing a `mvn deploy` you don't have that issue? If that's the case, the `maven-release-plugin` itself might add `wagon-ssh` to the classpath ( http://maven.apache.org/maven-release/maven-release-plugin/dependencies.html ) and that's wrong.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I think I know what's happening here: maven-release-plugin depends on maven2 dependencies. These pull in the ssh-wagen as transitive dependency. That would explain why it is only happening with the maven-release-plugin and not with a normal deploy. [Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=danttran) can confirm the fix, but since ssh is not part of the maven3 distribution, it shouldn't have any impact.

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

`@Jason`, I can confirm that maven 3.2.2 can download file via ssh as long as user uses wagon-2.9 as extension ( note maven does not have wagon-ssh under its lib)

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

`@if` user deploys artifact with ssh extension  wagon-2.8, it would hang at downloading metadata file

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Ok, I will stage Wagon 2.9 and you can give it a whirl.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

I consider this fixed with Wagon 2.9.
